instance-scheduler/environments.go

```diff
@@ -0,0 +1,154 @@
+}
+
+// JSONFileContent represents the structure of each JSON file's content
+type JSONFileContent map[string]interface{}
+
+// FetchJSON fetches the JSON content from a given URL
+func FetchJSON(rawURL string) (JSONFileContent, error) {
+    resp, err := http.Get(rawURL)
+    if err != nil {
+        return nil, fmt.Errorf("failed to fetch JSON content: %w", err)
+    }
+    defer resp.Body.Close()
+
+    if resp.StatusCode != http.StatusOK {
+        return nil, fmt.Errorf("non-200 status code: %d", resp.StatusCode)
+    }
+
+    body, err := io.ReadAll(resp.Body)
+    if err != nil {
+        return nil, fmt.Errorf("failed to read response body: %w", err)
+    }
+
+    var content JSONFileContent
+    err = json.Unmarshal(body, &content)
+	  if err != nil {
+		  return nil, fmt.Errorf("failed to unmarshal JSON: %w", err)
+	  }
+
+    return content, nil
+}
+
+// fetches the environments JSON data from GitHub
+func fetchGitHubData(baseURL, repoOwner, repoName, branch, directory string) ([]byte, error) {
+
+    //u, err := url.Parse(path.Join(baseURL, repoOwner, repoName, "contents", directory))
+    u, err := url.Parse(fmt.Sprintf("%s/%s/%s/contents/%s", baseURL, repoOwner, repoName, directory))
+    if err != nil {
+        return nil, fmt.Errorf("failed to parse URL: %w", err)
+    }
+
+    query := u.Query()
+    query.Set("ref", branch)
+    u.RawQuery = query.Encode()
+
+    // Print the constructed URL for debugging
+    fmt.Println("Constructed URL:", u.String())
+
+    // Create a new HTTP request
+    req, err := http.NewRequest("GET", u.String(), nil)
+    if err != nil {
+        return nil, fmt.Errorf("failed to create HTTP request: %w", err)
+    }
+
+    // Set the User-Agent header (GitHub API requires a User-Agent header)
+    req.Header.Set("User-Agent", "Go-http-client")
+
+    // Perform the HTTP request
+    client := &http.Client{}
+    resp, err := client.Do(req)
+    if err != nil {
+        return nil, fmt.Errorf("failed to fetch directory listing: %w", err)
+    }
+    defer resp.Body.Close()
+
+    if resp.StatusCode != http.StatusOK {
+        return nil, fmt.Errorf("non-200 status code: %d", resp.StatusCode)
+    }
+
+    body, err := io.ReadAll(resp.Body)
+    if err != nil {
+        return nil, fmt.Errorf("failed to read response body: %w", err)
+    }
+
+    return body, nil
+}
+
+// processGitHubData processes the JSON data and returns a slice of GitHubFile
+func processGitHubData(body []byte) ([]GitHubFile, error) {
+    var files []GitHubFile
+    err := json.Unmarshal(body, &files)
+    if err != nil {
+        return nil, fmt.Errorf("failed to unmarshal JSON: %w", err)
+    }
+    return files, nil
+}
+
+
+// hasInstanceSchedulerSkip checks if the instance_scheduler_skip field exists and contains "true"
+func hasInstanceSchedulerSkip(content gjson.Result) bool {
+    skip := content.Get("instance_scheduler_skip")
+    if skip.Exists() {
+        for _, value := range skip.Array() {
```

Comment:
Probably worth updating the underlying object to be `instance_scheduler_skip: "true"` instead of `instance_scheduler_skip: ["true"]` - saves us having to loop through the values, handle edge cases and keeps the interface for the user a bit simpler 😁